app/assets/javascripts/ui_models/app_state/notes_state.ts

```diff
@@ -0,0 +1,213 @@
+      )
+    ) {
+      if (this.selectedNotes[uuid]) {
+        delete this.selectedNotes[uuid];
+      } else {
+        this.selectedNotes[uuid] = note;
+        this.lastSelectedNote = note;
+      }
+    } else if (this.io.activeModifiers.has(KeyboardModifier.Shift)) {
+      const notes = this.application.getDisplayableItems(
+        ContentType.Note
+      ) as SNNote[];
+      const lastSelectedNoteIndex = notes.findIndex(
+        (note) => note.uuid == this.lastSelectedNote?.uuid
+      );
+      const selectedNoteIndex = notes.findIndex((note) => note.uuid == uuid);
+      notes
+        .slice(lastSelectedNoteIndex, selectedNoteIndex + 1)
+        .forEach((note) => (this.selectedNotes[note.uuid] = note));
+    } else {
+      this.selectedNotes = {
+        [uuid]: note,
+      };
+      this.lastSelectedNote = note;
+      await this.openEditor(uuid);
+    }
+  }
+
+  private async openEditor(noteUuid: string): Promise<void> {
+    if (this.activeEditor?.note?.uuid === noteUuid) {
+      return;
+    }
+
+    const note = this.application.findItem(noteUuid) as SNNote | undefined;
+    if (!note) {
+      console.warn('Tried accessing a non-existant note of UUID ' + noteUuid);
+      return;
+    }
+
+    if (await this.application.authorizeNoteAccess(note)) {
+      if (!this.activeEditor) {
+        this.application.editorGroup.createEditor(noteUuid);
+      } else {
+        this.activeEditor.setNote(note);
+      }
+      await this.onActiveEditorChanged();
+
+      if (note.waitingForKey) {
+        this.application.presentKeyRecoveryWizard();
+      }
+    }
+  }
+
+  setContextMenuOpen(open: boolean): void {
+    this.contextMenuOpen = open;
+  }
+
+  setContextMenuPosition(position: { top: number; left: number }): void {
+    this.contextMenuPosition = position;
+  }
+
+  setHideSelectedNotePreviews(hide: boolean): void {
+    this.application.changeItems<NoteMutator>(
+      Object.keys(this.selectedNotes),
+      (mutator) => {
+        mutator.hidePreview = hide;
+      },
+      false
+    );
+  }
+
+  setLockSelectedNotes(lock: boolean): void {
+    this.application.changeItems<NoteMutator>(
+      Object.keys(this.selectedNotes),
+      (mutator) => {
+        mutator.locked = lock;
+      },
+      false
+    );
+  }
+
+  async setTrashSelectedNotes(
+    trashed: boolean,
+    trashButtonRef: RefObject<HTMLButtonElement>
+  ): Promise<void> {
+    if (
+      await confirmDialog({
+        title: Strings.trashNotesTitle,
+        text: Strings.trashNotesText,
+        confirmButtonStyle: 'danger',
+      })
+    ) {
+      this.application.changeItems<NoteMutator>(
+        Object.keys(this.selectedNotes),
+        (mutator) => {
+          mutator.trashed = trashed;
+        },
+        false
+      );
+      runInAction(() => {
+        this.selectedNotes = {};
+        this.contextMenuOpen = false;
+      });
+    } else {
+      trashButtonRef.current?.focus();
```

Comment:
I think it would be best to keep actual UI elements out of state objects or we may experience some strange behavior/memory leaks. Instead UI components can get the truth from state objects and decide what to do with themselves. What about returning a boolean? Or having `confirmDialog` get invoked in a component, since it's technically UI as well?

Reply:
I've removed locking blur altogether for now since it wasn't working as expected. Looks like sometimes the blur event occurs before the lock state is actually updated, which means the menu gets closed anyway. Will have to take a further look to see if there's a workaround for it.